man/avg_showcamera.1
.TH AVG_SHOWCAMERA "1" "March 2011" "libavg 1.5.4" "User Commands"
.SH NAME
avg_showcamera \- camera discovery and display tool
.SH SYNOPSIS
.B avg_showcamera
\fI\-l\fR
.HP
.B avg_showcamera
\fI\-r\fR
.HP
.B avg_showcamera
\fI\-t DRIVER [OPTION]...\fR
.SH DESCRIPTION
Query and list attached cameras, reset firewire bus, show video stream.
.PP
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Show a brief description of the options.
.TP
\fB\-l\fR, \fB\-\-dump\fR
Dump a list of available cameras to the console.
.TP
\fB\-r\fR, \fB\-\-resetbus\fR
Reset the firewire bus and free allocated bandwidth.
.TP
\fB\-t\fR, \fB\-\-driver\fR DRIVER
One of the camera subsystems supported by libavg (supported: firewire, video4linux, directshow).
.TP
\fB\-d\fR, \fB\-\-device\fR DEVICE
Camera device identifier (depending on the driver, a GUID or device path is expected).
.TP
\fB\-u\fR, \fB\-\-unit\fR UNIT
Unit number. Used for cameras or other capture devices which deliver several images (e.g. stereo cameras or TV capture cards).
.TP
\fB\-w\fR, \fB\-\-width\fR WIDTH
Camera image width.
.TP
\fB\-e\fR, \fB\-\-height\fR HEIGHT
Camera image height.
.TP
\fB\-p\fR, \fB\-\-pixelformat\fR PIXELFORMAT
Camera pixel format (one of: I8, I16, YUV411, YUV422, YUYV422, RGB, BGR, BAYER8).
.TP
\fB\-f\fR, \fB\-\-framerate\fR FRAMERATE
Frames per second.
.TP
\fB\-8\fR, \fB\-\-fw800\fR
Set firewire bus speed to s800 (if supported).
.TP
\fB\-s\fR, \fB\-\-noinfo\fR
Don't show any info overlayed on the screen.
.SH AUTHOR
Ulrich von Zadow <coder@c-base.org>


NEWS
libavg 1.5.4 - February 28, 2011 (Linux only)

  svn revision 5644, branches/release1.5.x

  XInput 2.1 updates/fixes.

libavg 1.5.3 - February 20, 2011 (Linux only)

  svn revision 5599

  Major new feature:
    * Linux XInput 2.1 Multitouch support.

libavg 1.5.1 - November 14, 2010 (Mac only)

  svn revision 5275

  Fixed bug that caused fullscreen apps to crash on the mac (Bug # 162).

libavg 1.5.0 - November 6, 2010

  svn revision 5270

  Major New Features:
    * Nodes can be constructed directly from python, making createNode unnecessary.
    * Added support for rendering avg scenes to offscreen canvases.
    * Added GPU-based effects framework and Blur, Shadow, Chromakey effects.
    * Added multitouch driver model and support for Apple Magic Trackpads and TUIO 
      devices.
    * Support for flash videos with an alpha channel. Decoding is GPU-accellerated.

  Other Changes:
    * Added color controls (gamma, brightness, contrast) to all RasterNodes.
    * Node constructors have a parent parameter which makes a following addChild 
      redundant.
    * Added Player.setEventHook().
    * Added onscreen keyboard for touchscreens.
    * WordsNode.getNumLines(), .getCharIndexFromPos(), .getTextAsDisplayed().
    * DivNode.getEffectiveMediaDir().
    * Clearer profiling that works for multiple threads.
    * Added AVG_BREAK_ON_IMPORT and AVG_BREAK_ON_ASSERT env variables.
    * Added AVG_DUMP_TEST_FRAMES env variable.
    * Added Player.setWindowFrame().
    * Added VideoNode.queuelength.
    * Video stability & timing improvements.
    * DivNode crop default is now false.
    * Node class names are unified: All names end with 'Node': WordsNode, DivNode, etc.
    * Added ImageNode.compression for 16 bpp texture support.
    * Added avg_audioplayer.py.

    * Various compiler compatibility fixes: Compiles with gcc 4.5 and VC++ 2010 now.
    * Lots of bugfixes - see the bug tracker.

libavg 1.5.0.pre3 - November 4, 2010

  svn revision 5267

libavg 1.5.0.pre2 - November 3, 2010

  svn revision 5263

libavg 1.5.0.pre1 - October 31, 2010

  svn revision 5244 

libavg 1.0.0 - January 8, 2010

  svn revision 4489

  New Features:

    * hinting is now optional in words node.
    * Multitouch emulation module added to AVGMTApp.
    * masks now work with words nodes.
    * Added video.volume property and video.hasAudio() method.
    * Added AVG_LOG_CATEGORIES environment variable. Log configuration is now
      done using the environment and not by calling a method.
    * Added maskpos and masksize attribute so masks can be moved and rescaled.
    * Added Player.getKeyModifierState()
    * Added sse2 version of yuv->rgb color conversion.
    * Added avg_videoinfo.sh.
    * Added mesh node.
    * Added avg_showcamera --resetbus to reset the firewire bus after crashes.

  Other Improvements:
    * New anim framework is now stable.
    * Snow Leopard compatibility.
    * avg_showcamera has much better command line handling.
    * The test suite is faster and a lot more comprehensive.
    * hrefs and mediadirs are now unicode-aware, so filenames like "ö.png" 
      should work.
    * node.unlink(True) now deletes all event handlers in the node, making
      it a lot easier to avoid memory leaks. 
    * Lots of bug fixes:
      - Windows gamma is now reset correctly on program end.
      - Linux vsync now much more stable.
      - Setting node.pos.x now throws an error instead of being ignored. 
        Same for all other Point2D attributes.
      - video.getNumFrames() now returns the exact number of frames.
      - Behaviour of nodes before Player.play() and when they are not in the
        avg tree is now much more stable.
      - Fixed words node sensitive area for centered and right-aligned nodes.

libavg 1.0.0.pre3 - January 6, 2010

  svn revision 4475

libavg 1.0.0.pre2 - December 23, 2009

  svn revision 4398

libavg 1.0.0.pre1 - December 21, 2009

  svn revision 4387

libavg 0.9.0 - August 23, 2009

  svn revision 4100

  Major new features:

    * Support for vector node types: Line, PolyLine, Curve, Circle, Polygon,
      Rectangle. Vector nodes can be textured, with the closed vectors taking two
      textures: one for the outline, one for the fill. Vector drawing is GPU-based.
    * Plugin support. Plugins written in C++ can define complete new node types.
    * Powerful (but still experimental) new animation framework.
    * Unified & much more stable camera drivers. Consistent support for all 
      resolutions & framerates we could test. Support for more than one camera on all
      platforms. fw800 is supported.
    * Uniform support for aggregate types in node interfaces (node.pos, node.pivot, 
      ...)

  Other improvements:
    * gcc 4.4 and ffmpeg 0.5.0 compatibility.
    * Added letterspacing support to words node.
    * Threaded videos are now the default.
    * Improved video seek and loop performance.
    * New Node::getElementByPos() method that returns the node at that point on the
      screen.
    * Added avg.svnrevision and avg.svndate properties so the avg version can be 
      queried.
    * New avg_showfont.py utility to help font selection.
    * New avg_showcamera.py utility to help camera setup.
    * New avg.getMemoryUsage function that returns the resident set size.
    * New Words.addFontDir().
    * Added support for mipmapping node textures, including videos.
    * rasternode.maskhref to support static alpha masks for any videos, word nodes,
      etc.

  Syntax and semantic changes:
    * In the course of adding aggregate types for node attributes, several attributes
      have changed:
      - words.size is now called words.fontsize.
      - pivotx and pivoty have been replaced by pivot.
    * right-aligned and centered words nodes are positioned differently.
    * avgtrackerrc syntax has changed significantly; see src/avgtrackerrc.minimal.
    * Support for firewire cameras using libdc1394 ver. 1 has been phased out. Please 

src/utils/avg_showcamera.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# libavg - Media Playback Engine.
# Copyright (C) 2003-2008 Ulrich von Zadow
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Current versions can be found at www.libavg.de
#

import optparse
import time
from libavg import avg
from libavg import parsecamargs

def checkCamera():
    if not(camNode.isAvailable()):
        Log.trace(Log.APP, "Could not open camera")
        exit(1)


def onKey(event):
    def addWhitebalance(du = 0, dv = 0):
        camNode.setWhitebalance(camNode.getWhitebalanceU() + du, 
                                camNode.getWhitebalanceV() + dv)
        print "u:", camNode.getWhitebalanceU(), "v:", camNode.getWhitebalanceV()
    
    def addGain(gain):
        camNode.gain += gain
        print "gain:", camNode.gain
    
    def addShutter(shutter):
        camNode.shutter += shutter
        print "shutter:", camNode.shutter
        
    if event.keystring == "w":
        print "Setting Whitebalance"
        camNode.doOneShotWhitebalance()
    
    elif event.keystring == "1":
        addWhitebalance(du = -1)
        
    elif event.keystring == "2":
        addWhitebalance(du = 1)
        
    elif event.keystring == "3":
        addWhitebalance(dv = -1)
        
    elif event.keystring == "4":
        addWhitebalance(dv = 1)
        
    elif event.keystring == "s":
        print "Saving camera image to camimage.png" 
        camNode.getBitmap().save("camimage.png")
    
    elif event.keystring == "left":
       addShutter(shutter = -1)
        
    elif event.keystring == "right":
        addShutter(shutter = 1)
    
    elif event.keystring == "up":
        addGain(gain = 1)
    
    elif event.keystring == "down":
        addGain(gain = -1)
    

parser = optparse.OptionParser()
parsecamargs.addOptions(parser)
parser.add_option("-l", "--dump", dest="dump", action="store_true", default=False,
          help="dump a list of detected cameras")
parser.add_option("-s", "--noinfo", dest="noinfo", action="store_true", default=False,
          help="don't show any info overlayed on the screen")
parser.add_option("-r", "--resetbus", dest="resetbus", action="store_true", default=False,
          help="reset the firewire bus.")

(options, args) = parser.parse_args()

if options.driver is None and not options.dump and not options.resetbus:
    parser.print_help()
    print
    print "Keys available when image is being displayed:"
    print "  w: Execute whitebalance."
    print "  1/2: Decrease/Increase whitebalance u."
    print "  3/4: Decrease/Increase whitebalance v."
    print
    print "ERROR: at least '--driver', '--dump' or '--resetbus' options must be specified"
    exit()

optdict = {}
for attr in dir(options):
    if attr[0] != '_':
        optdict[attr] = eval("options.%s" %attr)

Log = avg.Logger.get()

Player = avg.Player.get()

Player.loadString("""
<?xml version="1.0"?>
<!DOCTYPE avg SYSTEM "../../doc/avg.dtd">
<avg width="%(width)d" height="%(height)d">
</avg>
""" %optdict)
Player.getRootNode().setEventHandler(avg.KEYDOWN, avg.NONE, onKey)

if options.dump:
    avg.CameraNode.dumpCameras()
    exit(0)

if options.resetbus:
    Log.trace(Log.APP, "Resetting firewire bus.")
    avg.CameraNode.resetFirewireBus()
    time.sleep(1)
    if not options.driver:
        exit(0)

Log.trace(Log.APP, "Creating camera:")
Log.trace(Log.APP, "driver=%(driver)s device=%(device)s" %optdict)
Log.trace(Log.APP, "width=%(width)d height=%(height)d pixelformat=%(pixelFormat)s" 
        %optdict)
Log.trace(Log.APP, "unit=%(unit)d framerate=%(framerate)d fw800=%(fw800)s" %optdict)
    
camNode = Player.createNode("camera", 
        {"driver": options.driver, "device": options.device, "unit": options.unit, 
         "fw800": options.fw800, 
         "capturewidth": options.width, "captureheight": options.height, 
         "pixelformat": options.pixelFormat, "framerate": options.framerate, 
         "width": options.width, "height": options.height})

Player.getRootNode().appendChild(camNode)

if not options.noinfo:
    infoText = "Driver=%(driver)s (dev=%(device)s unit=%(unit)d) %(width)dx%(height)d@%(framerate)f" %optdict
    infoNode = Player.createNode("words",
    {"text": infoText, "color": "ff3333", "pos": avg.Point2D(5,5), "fontsize": 14})

    Player.getRootNode().appendChild(infoNode)
    

camNode.play()
Player.setTimeout(100, checkCamera)
Player.play()

